File: homework3/Squares/Squares.cs
using System;
using System.Threading;

class Squares
{
    private static int sum = 0;
    private static readonly object lockObject = new object();

    static void Main(string[] args)
    {
        int start = 1;
        int end = 100;
        int threadCount = 4; // Number of threads to use
        int range = (end - start + 1) / threadCount;

        Thread[] threads = new Thread[threadCount];

        for (int i = 0; i < threadCount; i++)
        {
            int threadStart = start + i * range;
            int threadEnd = (i == threadCount - 1) ? end : threadStart + range - 1;

            threads[i] = new Thread(() => CalculateSquares(threadStart, threadEnd));
            threads[i].Start();
        }

        foreach (Thread thread in threads)
        {
            thread.Join();
        }

        Console.WriteLine($"The sum of squares from {start} to {end} is: {sum}");
    }

    private static void CalculateSquares(int start, int end)
    {
        int localSum = 0;
        for (int i = start; i <= end; i++)
        {
            localSum += i * i;
        }

        lock (lockObject)
        {
            sum += localSum;
        }
    }
}

File: homework3/Squares/Squares.kt
import kotlin.concurrent.thread

fun main() {
    var sum = 0L // 공유 자원
    val lock = Any() // Lock 객체
    val threads = mutableListOf<Thread>()
    val range = 1..100
    val numThreads = 4 // Number of threads
    val chunkSize = (range.last - range.first + 1) / numThreads

    for (i in 0 until numThreads) {
        val start = range.first + i * chunkSize
        val end = if (i == numThreads - 1) range.last else start + chunkSize - 1
        threads.add(thread {
            val localSum = (start..end).sumOf { it.toLong() * it }
            synchronized(lock) { // 크리티컬 섹션
                sum += localSum
            }
        })
    }

    threads.forEach { it.join() } // Wait for all threads to finish
    println("The sum of squares from ${range.first} to ${range.last} is: $sum")
}